main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -54,7 +54,240 @@ object OccurrenceAnalyzer {
   /**
     * Performs occurrence analysis on `exp0`
     */
-  private def visitExp(exp0: OccurrenceAst.Expr)(implicit sym0: Symbol.DefnSym, lctx: LocalContext): (OccurrenceAst.Expr, ExpContext) = (exp0, ExpContext.empty)
+  private def visitExp(exp0: OccurrenceAst.Expr)(implicit sym0: Symbol.DefnSym, lctx: LocalContext): (OccurrenceAst.Expr, ExpContext) = exp0 match {
+    case OccurrenceAst.Expr.Cst(cst, tpe, loc) =>
+      lctx.size.incrementAndGet()
+      (OccurrenceAst.Expr.Cst(cst, tpe, loc), ExpContext.empty)
+
+    case OccurrenceAst.Expr.Var(sym, tpe, loc) =>
+      (OccurrenceAst.Expr.Var(sym, tpe, loc), ExpContext.empty.addVar(sym, Once))
+
+    case OccurrenceAst.Expr.Lambda(fp, exp, tpe, loc) =>
+      val (e, ctx1) = visitExp(exp)
+      val ctx2 = ctx1.map {
+        case Once => OnceInLambda
+      }
+      val occur = ctx2.get(fp.sym)
+      val ctx3 = ctx2.removeVar(fp.sym)
+      lctx.size.incrementAndGet()
+      (OccurrenceAst.Expr.Lambda(fp.copy(occur = occur), e, tpe, loc), ctx3)
+
+    case OccurrenceAst.Expr.ApplyAtomic(op, exps, tpe, eff, loc) =>
+      val (es, ctxs) = exps.map(visitExp).unzip
+      val ctx = ctxs.foldLeft(ExpContext.empty)(combineSeq)
+      lctx.size.incrementAndGet()
+      (OccurrenceAst.Expr.ApplyAtomic(op, es, tpe, eff, loc), ctx)
+
+    case OccurrenceAst.Expr.ApplyClo(exp1, exp2, tpe, eff, loc) =>
+      val (e1, ctx1) = visitExp(exp1)
+      val (e2, ctx2) = visitExp(exp2)
+      val ctx3 = combineSeq(ctx1, ctx2)
+      lctx.size.incrementAndGet()
+      (OccurrenceAst.Expr.ApplyClo(e1, e2, tpe, eff, loc), ctx3)
```

Comment:
We can make an important optimization here. It looks like this:

```
if (e1 eq exp1 && e2 eq exp2){
  (exp0, ...) // reuse memory
else 
  (...ApplyClo, ...)
```

Reply:
Because the subtrees are big?

Reply:
Yes